asserts/validation_set.go

```diff
@@ -0,0 +1,262 @@
+		SnapID:   snapID,
+		Presence: presence,
+		Revision: snapRevision,
+	}, nil
+}
+
+func checkValidationSetSnaps(snapList interface{}) ([]*ValidationSetSnap, error) {
+	const wrongHeaderType = `"snaps" header must be a list of maps`
+
+	entries, ok := snapList.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(wrongHeaderType)
+	}
+
+	seen := make(map[string]bool, len(entries))
+	seenIDs := make(map[string]string, len(entries))
+	snaps := make([]*ValidationSetSnap, 0, len(entries))
+	for _, entry := range entries {
+		snap, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(wrongHeaderType)
+		}
+		valSetSnap, err := checkValidationSetSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+
+		if seen[valSetSnap.Name] {
+			return nil, fmt.Errorf("cannot list the same snap %q multiple times", valSetSnap.Name)
+		}
+		seen[valSetSnap.Name] = true
+		snapID := valSetSnap.SnapID
+		if underName := seenIDs[snapID]; underName != "" {
+			return nil, fmt.Errorf("cannot specify the same snap id %q multiple times, specified for snaps %q and %q", snapID, underName, valSetSnap.Name)
+		}
+		seenIDs[snapID] = valSetSnap.Name
+
+		if valSetSnap.Presence == "" {
+			valSetSnap.Presence = PresenceRequired
+		}
+
+		snaps = append(snaps, valSetSnap)
+	}
+
+	return snaps, nil
+}
+
+// ValidationSet holds a validation-set assertion, which is a
+// statement by an account about a set snaps and possibly revisions
+// for which an extrisinc/implied property is valid (e.g. they work
+// well together). validation-sets are organized in sequences under a
+// name.
+type ValidationSet struct {
+	assertionBase
+
+	seq int
+
+	snaps []*ValidationSetSnap
+
+	timestamp time.Time
+}
+
+// Series returns the series for which the snap in the set are declared.
+func (vs *ValidationSet) Series() string {
+	return vs.HeaderString("series")
+}
+
+// AccountID returns the identifier of the account that signed this assertion.
+func (vs *ValidationSet) AccountID() string {
+	return vs.HeaderString("account-id")
+}
+
+// Name returns the name under which the validation-set is organized.
+func (vs *ValidationSet) Name() string {
+	return vs.HeaderString("name")
+}
+
+// Sequence returns the sequential number of the validation-set in its
+// named sequence.
+func (vs *ValidationSet) Sequence() int {
+	return vs.seq
+}
+
+// Snaps returns the constrained snaps by the validation-set.
+func (vs *ValidationSet) Snaps() []*ValidationSetSnap {
+	return vs.snaps
+}
+
+// Timestamp returns the time when the validation-set was issued.
+func (vs *ValidationSet) Timestamp() time.Time {
+	return vs.timestamp
+}
+
+func checkSequence(headers map[string]interface{}, name string) (int, error) {
+	seqnum, err := checkInt(headers, name)
+	if err != nil {
+		return -1, err
+	}
+	if seqnum < 1 {
+		return -1, fmt.Errorf("%q must be >=1: %v", name, seqnum)
+	}
+	return seqnum, nil
+}
+
+var (
+	validValidationSetName = regexp.MustCompile("^[a-z0-9](?:-?[a-z0-9])*$")
```

Comment:
If I read this correctly, a single-character name is OK, but I don't think we have a test case for that below, perhaps it's worth adding one.

Reply:
we don't tend to have such positive tests, maybe we should have some though for model and validation-set names

Reply:
FWTW we have such tests for snap names (that's where I'm coming from after recent bug concerning 1-character names ;)).
Not super ciritical for this PR but nice to have to make sure we don't accidentally replace "*" with "+" in the future.